core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -26,21 +26,44 @@ import cats.data.{Chain, NonEmptyList}
 
 import cats.implicits._
 
-/** Following the haskell library trifecta,
-  * these parsers either parse successfully, parse 0 and fail,
-  * or parse more than 1 and fail. The "or" operation
-  * only works for parsing 0 and failing.
-  *
-  * We call a failure without parsing anything an epsilon failure.
-  *
-  * We can convert a failure after parsing more than one
-  * to a zero parse with .backtrack. Related to backtracking
-  * is softProduct, which if the second failure is an espilon
-  * failure, yields an epsilon failure for the product.
-  *
-  * Parses 0 or more characters to return A
+/**
+ * Parser[A] attempts to extract an `A` value from the given input,
+ * potentially moving its offset forward in the process.
+ *
+ * When calling `parse`, one of three outcomes occurs:
+ *
+ *   - Success: The parser consumes zero-or-more characters of input
+ *     and successfully extracts a value. The input offset will be
+ *     moved forward by the number of characters consumed.
+ *
+ *   - Epsilon failure: The parser fails to extract a value without
+ *     consuming any characters of input. The input offset will not be
+ *     changed.
+ *
+ *   - Arresting failure: The parser fails to extract a value but does
+ *     consume one-or-more characters of input. The input offset will
+ *     be moved forward by the number of characters consumed. The
+ *     consumed characters will not be available to any additional
+ *     parsers which may run.
+ *
+ * Operations such as `x.orElse(y)` will only consider parser `y` if
+ * `x` returns an epsilon failure; these methods cannot recover from
+ * an arresting failure. Arresting failures can be "rewound" using
+ * methods such as `backtrack` or `softProduct`, which converts
+ * arresting failures into epsilon failures. Since rewinding is
```

Comment:
Rewinding is actually cheap, but backtracking can result in exponential complexity and tends to make errors harder to locate.